### PR TITLE
UX: Add progress bar for both Git and Mercurial when using non-verbose mode

### PR DESCRIPTION
**Pull request by screamerbg**
**This PR adds**
* Progress bar for both Git and Mercurial when cloning

The progress bar is only visible when mbed CLI is in non-verbose mode (no `-v` or `-vv`).

**Git example:**
```
/tmp:$ mbed import mbed-os-example-client
[mbed] Importing program "mbed-os-example-client" from "https://github.com/ARMmbed/mbed-os-example-client" at latest revision in the current branch
[mbed] Adding library "easy-connect" from "https://github.com/ARMmbed/easy-connect" at rev #6af200df4817
[mbed] Adding library "easy-connect/atmel-rf-driver" from "https://github.com/ARMmbed/atmel-rf-driver" at rev #2cd9abba3e37
[mbed] Adding library "easy-connect/esp8266-driver" from "https://github.com/ARMmbed/esp8266-driver" at rev #b0d79dad507d
[mbed] Adding library "easy-connect/mcr20a-rf-driver" from "https://github.com/ARMmbed/mcr20a-rf-driver" at rev #9aac10474702
[mbed] Adding library "easy-connect/stm-spirit1-rf-driver" from "https://github.com/ARMmbed/stm-spirit1-rf-driver" at rev #0ff4ca7537f0
[mbed] Adding library "easy-connect/wifi-x-nucleo-idw01m1" from "https://github.com/ARMmbed/wifi-x-nucleo-idw01m1" at rev #537b01c4a053
[mbed] Adding library "mbed-client" from "https://github.com/ARMmbed/mbed-client" at rev #ea04c5de7822
[mbed] Adding library "mbed-client/mbed-client-c" from "https://github.com/ARMmbed/mbed-client-c" at rev #ecfa619e42b2
[mbed] Adding library "mbed-client/mbed-client-classic" from "https://github.com/ARMmbed/mbed-client-classic" at rev #4e66929607c3
[mbed] Adding library "mbed-client/mbed-client-mbed-tls" from "https://github.com/ARMmbed/mbed-client-mbed-tls" at rev #7e1b6d815038
[mbed] Adding library "mbed-os" from "https://github.com/ARMmbed/mbed-os" at rev #8b54959f5bca
Downloading |##########------------------------------------------------| 16.17%
```

**Hg example:**
```
/tmp/1:$ mbed import https://os.mbed.com/users/mbed_official/code/mbed-dev
[mbed] Importing program "mbed-dev" from "https://os.mbed.com/users/mbed_official/code/mbed-dev" at latest revision in the current branch
Downloading |######################------------------------------------| 36.75%
```

Note that with Mercurial/Hg the progress bar appears late because of the progress reported late by Mercurial.

**Backwards compatbility**
This PR is fully backwards compatible

**Documentation**
No documentation is needed

### CODE REVIEW

File: mbed/mbed.py [hunk 1]
```diff
@@ -177,6 +177,20 @@ def progress():
...
+    bwidth = max_width - len(str(title)) - len(show_percent) - 6
+    for i in range(0, bwidth):
+        line += '#' if i <= (percent / 100 * bwidth) else '-'
+    line += '| ' + show_percent + '%'
+    sys.stdout.write(line+'\r')
+    sys.stdout.flush()
+    sys.stdout.write('\b')
+
+def hide_progress():
+    sys.stdout.write("\033[K\r\b")
```
Comment (theotherjimmy) [inline mbed/mbed.py]: This seems a little magic. Does this work cross platform?

Comment (screamerbg) [reply mbed/mbed.py]: I've checked it on Mac OS, Linux, Windows 7 and Windows 10.
Esc[K means Clear line from cursor right

Escape sequences are standardized for years for ANSI/VT-100 compatible terminals. http://ascii-table.com/ansi-escape-sequences-vt-100.php.

Not sure how this would look in a Jenkins log though as it's not a terminal. Any ideas how we could test with Jenkins?

Comment (screamerbg) [reply mbed/mbed.py]: Just noticed that on Windows there's a parasite character. Pending minor patch to solve it.

Comment (screamerbg) [reply mbed/mbed.py]: @theotherjimmy Magic removed in favour of more standard cursor handling :)

File: mbed/mbed.py [hunk 2]
```diff
@@ -177,6 +177,22 @@ def progress():
     sys.stdout.flush()
     sys.stdout.write('\b')
 
+def show_progress(title, percent, max_width=80):
+    percent = round(float(percent), 2)
+    show_percent = '%.2f' % percent
+    line = str(title) + ' |'
+    bwidth = max_width - len(str(title)) - len(show_percent) - 6
```
Comment (theotherjimmy) [inline mbed/mbed.py]: This `6` seems a litte magic. Is it `len(' |') + len('| ') + len('%') + len('\r')`?

File: mbed/mbed.py [hunk 3]
```diff
@@ -177,6 +177,22 @@ def progress():
...
+    for i in range(0, bwidth):
+        line += '#' if i <= (percent / 100 * bwidth) else '-'
+    line += '| ' + show_percent + '%'
+    sys.stdout.write(line+'\r')
+    sys.stdout.flush()
+
+def hide_progress(max_width=80):
+    line = ''
+    for i in range(0, max_width):
+        line += ' '
```
Comment (theotherjimmy) [inline mbed/mbed.py]: These 3 lines could be:
```python
line = ' ' * max_width
```
Up to you. I have a preference for the above because I think it better conveys the intent.

File: mbed/mbed.py [hunk 4]
```diff
@@ -211,7 +227,21 @@ def pquery(command, stdin=None, **kwargs):
...
+            if s == '\r' or s == '\n':
+                output_callback(line, s)
+                line = ""
+
+            if proc.returncode is None:
+                code = proc.poll()
+            else:
+                break
+
+    stdout, stderr = proc.communicate(stdin)
```
Comment (theotherjimmy) [inline mbed/mbed.py]: I don't think we started using `stderr` here, so it may be best to revert this change.

File: mbed/mbed.py (hunk 2 above)
Comment (screamerbg) [reply mbed/mbed.py]: It's the ones you mentioned + an extra character as on some terminals (like Windows) printing exactly 80 characters, produces a new line.

File: mbed/mbed.py (hunk 3 above)
Comment (screamerbg) [reply mbed/mbed.py]: Great that you spotted this. Will change it shortly.

File: mbed/mbed.py (hunk 4 above)
Comment (screamerbg) [reply mbed/mbed.py]: Reverted. Thanks
